sdk/cosmos/azure-cosmos/azure/cosmos/aio/container.py

```diff
@@ -0,0 +1,742 @@
+        doc_link = self._get_document_link(item)
+        request_options = build_options(kwargs)
+        response_hook = kwargs.pop('response_hook', None)
+        if partition_key is not None:
+            request_options["partitionKey"] = await self._set_partition_key(partition_key)
+        if populate_query_metrics is not None:
+            request_options["populateQueryMetrics"] = populate_query_metrics
+        if post_trigger_include is not None:
+            request_options["postTriggerInclude"] = post_trigger_include
+
+        result = await self.client_connection.ReadItem(document_link=doc_link, options=request_options, **kwargs)
+        if response_hook:
+            response_hook(self.client_connection.last_response_headers, result)
+        return result
+
+    @distributed_trace
+    def read_all_items(
+        self,
+        max_item_count=None,  # type: Optional[int]
+        populate_query_metrics=None,  # type: Optional[bool]
+        **kwargs  # type: Any
+    ):
+        # type: (...) -> Iterable[Dict[str, Any]]
+        """List all the items in the container.
+
+        :param max_item_count: Max number of items to be returned in the enumeration operation.
+        :param populate_query_metrics: Enable returning query metrics in response headers.
+        :keyword str session_token: Token for use with Session consistency.
+        :keyword dict[str,str] initial_headers: Initial headers to be sent as part of the request.
+        :keyword Callable response_hook: A callable invoked with the response metadata.
+        :returns: An Iterable of items (dicts).
+        :rtype: Iterable[dict[str, Any]]
+        """
+        feed_options = build_options(kwargs)
+        response_hook = kwargs.pop('response_hook', None)
+        if max_item_count is not None:
+            feed_options["maxItemCount"] = max_item_count
+        if populate_query_metrics is not None:
+            feed_options["populateQueryMetrics"] = populate_query_metrics
+
+        if hasattr(response_hook, "clear"):
+            response_hook.clear()
+
+        items = self.client_connection.ReadItems(
+            collection_link=self.container_link, feed_options=feed_options, response_hook=response_hook, **kwargs
+        )
+        if response_hook:
+            response_hook(self.client_connection.last_response_headers, items)
+        return items
+
+    @distributed_trace_async
+    async def query_items(
+        self,
+        query,  # type: str
+        parameters=None,  # type: Optional[List[Dict[str, object]]]
+        partition_key=None,  # type: Optional[Any]
+        enable_cross_partition_query=None,  # type: Optional[bool]
+        max_item_count=None,  # type: Optional[int]
+        enable_scan_in_query=None,  # type: Optional[bool]
+        populate_query_metrics=None,  # type: Optional[bool]
+        **kwargs  # type: Any
+    ):
+        # type: (...) -> Iterable[Dict[str, Any]]
+        """Return all results matching the given `query`.
```

Comment:
I can't find any api for returning response headers other than `self.client_connection.last_response_headers` for any of the APIs. 

what's the workaround for getting response headers, assuming that's something python users would need?

Reply:
that's all the sdk currently has as of now - once the response is returned after the function gets sent into _PipelineRunFunction, the last response headers get returned along with the response body

Reply:
the last header response is not well defined when iterating through the results of a query, also it is not possible to associate the items in the result to the response header (there is no clear boundary).

@simorenoh could you check which pattern the other azure python SDKs are using for this?